Sam361 update urdb api

cpaulgilman opened the pull request:
Update to URDB Version 7

Some utilities like SCE have more than 500 rates, so API limit on number of returned rates prevents all rates. Implement Jay Huggins suggestion:

> You can include the “offset” param in your calls, starting with 0, and increment it by the limit value and repeat the calls until you get < 500 results (wouldn’t hurt to explicitly include the limit of 500 for clarity), and page through all results, no matter how many there are.  

Improve layout and messagingn in URDB Download window

Get metering / billing option if available

Correctly handle minimum charges: `minmonthlycharge` and `annualmincharge` replaced my `mincharge` and `minchargeunits` per https://openei.org/services/doc/rest/util_rates/?version=7.

Calculate fixed charge from `fixedchargefirstmeter`, `fixedchargeeaaddl`, and `fixedchargeunits` 

For items SAM does not model, get as "unused" item:

* `fueladjustmentsmonthly` $/kWh

* `demandwindow`

* `coincidentrateschedule`, `coincidentratestructure`, `coincidentrateunit`

* `demandratchetpercentage`

* `demandreactivepowercharge`

* `energyattrs`

* `fixedattrs`

* `is_default`

* `servicetype`

Add rate notes input for fixed and minimum charges in units that SAM does not handle.

The following items available from the URDB that SAM does not use are now better documented. We could add some of these in the future:

* minchargeunits = $/day

* fixedchargeunits = {$/day, $/year} (could convert these to $/month)

* dgRules Net Billing Hourly vs Net Billing Instantaneous, for now assign either to Net Billing.

* fueladjustmentsmonthly (monthly array of $/kWh rates)

* coincident demand charges

* demand ratchets

* demand reactive power charge

* demand units (applicability)

sjanzou reviewed (approved):
@cpaulgilman, Changes look great! Not sure how to advise merging since there are defaults and ui forms and lk text file updates and Github is poor at best at merging these. When did you last merge in develop? I would suggest leaving out the test results and merge into develop and then use the new test results. We will need to carefully review defaults and UI pages for all features desired from here and develop. The improvements here are well worth it! Let me know if I may help with the merge.

cpaulgilman commented:
I merged Develop into this branch and then ran the test script just before making the pull request. I agree this will be a tricky merge given the UI and defaults changes -- I will test and be prepared for possible manual updates in Develop after the merge.

Note that invoke.cpp and webapis.conf will need to be updated in sam-private as well.

brtietz commented:
I'm having some trouble getting this working in the open source version. Is that because this now relies on private.h within SAM-private for the API key?

cpaulgilman commented:
@brtietz Yes. To test any web service features, you have to temporarily copy the API keys from private.h in sam-private to sam and be careful not to commit private.h. Those keys should never have been committed to the open source version of private.h.

brtietz commented on src/invoke.cpp:
```diff
@@ -2872,20 +2906,62 @@ void fcall_urdb_get(lk::invoke_t &cxt)
+        if ((rate.FixedChargeUnits) == "$/month")
+            cxt.result().hash_item("monthly_fixed_charge").assign(fixed_charges);
+        else if ( fixed_charges > 0 )
+            rate_notes.append(wxString::Format("SAM does not model fixed charge rate of %f with %s units.\n", fixed_charges, rate.FixedChargeUnits));
```
I tried downloading this rate: https://openei.org/apps/IURDB/rate/view/5c6f06ff5457a37c10399c02#4__Fixed_Charges
which has fixed charges of "$/day" and did not get this message. Fixed charges did not appear in the "unused items" either.

brtietz commented on src/invoke.cpp:
```diff
@@ -2872,20 +2906,62 @@ void fcall_urdb_get(lk::invoke_t &cxt)
+		else if ( rate.MinChargeUnits == "$/year")
+			cxt.result().hash_item("annual_min_charge").assign(rate.MinCharge);
+        else if (rate.MinCharge > 0 )
+            rate_notes.append(wxString::Format("SAM does not model minimum charge rate of %f with %s units.\n", rate.MinCharge, rate.MinChargeUnits));
```
Similar problem with this rate: https://en.openei.org/apps/IURDB/rate/view/5c817fbf7bece2299eceb397 and the error message for minimum charges: https://en.openei.org/apps/IURDB/rate/view/5c817fbf7bece2299eceb397

brtietz reviewed (changes requested):
User notifications for $/day units are not showing up. See URLs for rates in comments on the individual lines.

brtietz commented:
For future reference/testing, I used the "conditions" box and some JSON on this page to find those rates:
![image](https://user-images.githubusercontent.com/5530592/98573710-628eab80-2274-11eb-99ee-936d60fd2445.png)

Thoughts on where to document this for future reference? This doc, perhaps? https://github.com/NREL/SAM-documentation/blob/master/Developer%20Tools%20and%20Instructions/URDB_Upload_Instructions.md

cpaulgilman commented:
Thanks @brtietz. Review items addressed in e41dccd76.

cpaulgilman commented:
Documentation for URDB downloads and uploads:

Added https://github.com/NREL/SAM-documentation/blob/master/Developer%20Tools%20and%20Instructions/urdb-download-urls-and-ui-notes-.md

Renamed https://github.com/NREL/SAM-documentation/blob/master/Developer%20Tools%20and%20Instructions/urdb-add-rate-instructions.md

brtietz reviewed (approved):
The update to rate notes looks great, thanks!

cpaulgilman commented:
Close #361
